File: src/Enterprise/Kotlin/FacadeExample.kt
class AuthService {
    fun authenticate(user: String) = "auth($user)"
}

class InventoryService {
    fun reserve(sku: String) = "reserve($sku)"
}

class BillingService {
    fun charge(cents: Int) = "charge($cents)"
}

class CheckoutFacade(
    private val auth: AuthService,
    private val inventory: InventoryService,
    private val billing: BillingService,
) {
    fun checkout(user: String, sku: String, cents: Int): String =
        "checkout=${auth.authenticate(user)}>${inventory.reserve(sku)}>${billing.charge(cents)}"
}

fun main() {
    val facade = CheckoutFacade(AuthService(), InventoryService(), BillingService())
    println(facade.checkout("alice", "SKU-42", 499))
}

File: src/Enterprise/VB.NET/FacadeExample.vb
Imports System

Friend NotInheritable Class AuthService
    Public Function Authenticate(user As String) As String
        Return $"auth({user})"
    End Function
End Class

Friend NotInheritable Class InventoryService
    Public Function Reserve(sku As String) As String
        Return $"reserve({sku})"
    End Function
End Class

Friend NotInheritable Class BillingService
    Public Function Charge(cents As Integer) As String
        Return $"charge({cents})"
    End Function
End Class

Friend NotInheritable Class CheckoutFacade
    Private ReadOnly auth As AuthService
    Private ReadOnly inventory As InventoryService
    Private ReadOnly billing As BillingService

    Public Sub New(auth As AuthService, inventory As InventoryService, billing As BillingService)
        Me.auth = auth
        Me.inventory = inventory
        Me.billing = billing
    End Sub

    Public Function Checkout(user As String, sku As String, cents As Integer) As String
        Return $"checkout={auth.Authenticate(user)}>{inventory.Reserve(sku)}>{billing.Charge(cents)}"
    End Function
End Class

Module FacadeExample
    Sub Main()
        Dim facade = New CheckoutFacade(New AuthService(), New InventoryService(), New BillingService())
        Console.WriteLine(facade.Checkout("alice", "SKU-42", 499))
    End Sub
End Module
